### ISSUE TITLE
Allow custom router-level middleware

### ISSUE THREAD

**Issue by b055man** [davesag/swagger-routes-express#22]
Currently the connector wraps the whole router-level logic and allows only minor changes to it - by means of the security middleware.
However, it'd be great if you enable option for passing an arbitrary middleware that requires to work on a router-level (https://expressjs.com/en/guide/using-middleware.html#middleware.router) - as it is in case of e.g. express-ajv-swagger-validation module: https://www.npmjs.com/package/express-ajv-swagger-validation

From the connector perspective, it can be as easy as adding a single line:
      if (opts['customMiddlewares']) descriptor.push(opts['customMiddlewares'])

**Comment by davesag** [comment]
Interesting suggestion @b055man — I'll take a look over the weekend.

**Comment by davesag** [comment]
Having looked into this I believe you can achieve what you want either by just creating the `router` object for a specific `api.yml` file and passing that into the `connect` function in your server startup script. The downside to that is you need multiple `api.yml` files for each part of the router. I might need to make some code changes to allow use of multiple files however.

The other alternative is to add a `'x-use-router'` variable in the `api.yml` definition and let It manage the creation of router objects based on the settings in that.

I'd rather you considered the first option as a fill-in measure while I look further at all the implications of the second option.

**Comment by davesag** [comment]
Hi @b055man 

I believe I misunderstood your request now I've looked at this.  Are you after the ability to define custom middleware on a route by route basis?

In which case I propose supporting an [OpenAPI Specification Extension](https://swagger.io/specification/#specificationExtensions) in the form `x-middleware` as follows

```yml
paths:
  /some-route:
    get:
      summary: 'some summary'
      security:
        - apiKey: []
      x-middleware:
        - customMiddleware
```

Then the descriptor would end up being
```js
[path, apiKeyMiddleware, customMiddlewareHandler, handler]
```

You'd need to pass in the `middleware` via the options:

```js
const options = {
  middleware: {
    customMiddleware: customMiddlewareHandler
  }
}
```

Would that suit your needs?

**Comment by b055man** [comment]
Hi @davesag,

yes, that's something I was thinking about. One thing though: could customMiddleware be an array, with an ordered list of middlewares?

A.

**Comment by davesag** [comment]
Hey @b055man yes sure, though the order of the middleware would be the order defined in the OpenAPI.yml or swagger.yml not the options as the options are just there as a lookup.  For example if your api is defined as

```yml

paths:
  /some-route:
    get:
      summary: 'some summary'
      operationId: handleSomeRoute
      security:
        - apiKey: []
      x-middleware:
        - doThis
        - doThat
```

Then your options contain

```js
const options = {
  middleware: {
    doThis: thisDoerFn,
    doThat: thatDoerFn
  }
}
```

then the descriptor for the route will be

```js
['/some-route', thisDoerFn, thatDoerFn, handleSomeRoute]
```

**Comment by b055man** [comment]
Yes, that's understandable. 
So, I assume you are planning to introduce this functionality? 

Thanks,
A.

**Comment by davesag** [comment]
@b055man yep.  I'll add it as part of the Version 2.1 release. (It could probably be done before that if you have an urgent need for it).

**Comment by davesag** [comment]
Hey @b055man this is now done and was merged in with PR #33 

I am just waiting on the new release of [`mocha` with the fix for the security audit issue](https://github.com/mochajs/mocha/pull/3845) before I publish a new version.  I'll do that as soon as mocha is updated.

**Comment by davesag** [comment]
Hi @b055man this feature is now available in version 2.1.0.  Thanks for your suggestion.

**Comment by b055man** [comment]
@davesag thanks a lot. I'll give it a try after the weekend.
